From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Auke Kok <auke-jan.h.kok@intel.com>
Date: Wed, 7 Nov 2018 14:38:46 -0800
Subject: [PATCH] Stateless moduli.

[2021-04-19]: Rebased for update to 8.6p1; fixed error message to account for
              stateless fallpath path.

Signed-off-by: Patrick McCarty <patrick.mccarty@intel.com>
---
 dh.c        | 4 ++--
 pathnames.h | 2 ++
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/dh.c b/dh.c
index ce2eb47..30db1e2 100644
--- a/dh.c
+++ b/dh.c
@@ -164,9 +164,9 @@ choose_dh(int min, int wantbits, int max)
 	int best, bestcount, which, linenum;
 	struct dhgroup dhg;
 
-	if ((f = fopen(get_moduli_filename(), "r")) == NULL) {
+	if ((f = fopen(get_moduli_filename(), "r")) == NULL && (f = fopen(_PATH_DH_MODULI_STATELESS, "r")) == NULL) {
 		logit("WARNING: could not open %s (%s), using fixed modulus",
-		    get_moduli_filename(), strerror(errno));
+		    _PATH_DH_MODULI_STATELESS, strerror(errno));
 		return (dh_new_group_fallback(max));
 	}
 
diff --git a/pathnames.h b/pathnames.h
index f7ca5a7..af527f2 100644
--- a/pathnames.h
+++ b/pathnames.h
@@ -42,6 +42,8 @@
 #define _PATH_HOST_XMSS_KEY_FILE	SSHDIR "/ssh_host_xmss_key"
 #define _PATH_HOST_RSA_KEY_FILE		SSHDIR "/ssh_host_rsa_key"
 #define _PATH_DH_MODULI			SSHDIR "/moduli"
+/* Stateless moduli fallback */
+#define _PATH_DH_MODULI_STATELESS      	"/usr/share/defaults/ssh/moduli"
 
 #ifndef _PATH_SSH_PROGRAM
 #define _PATH_SSH_PROGRAM		"/usr/bin/ssh"
